src/lib/device/virtual_dev.cpp

```diff
@@ -217,41 +211,66 @@ BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints
     }
 }
 
-BlkAllocStatus VirtualDev::alloc_blk_from_chunk(blk_count_t nblks, const blk_alloc_hints& hints,
-                                                std::vector< BlkId >& out_blkid, Chunk* chunk) {
+BlkAllocStatus VirtualDev::alloc_blks(blk_count_t nblks, blk_alloc_hints const& hints,
+                                      std::vector< BlkId >& out_blkids) {
+    // Regular alloc blks will allocate in MultiBlkId, but there is an upper limit on how many it can accomodate in a
+    // single MultiBlkId, if caller is ok to generate multiple MultiBlkids, this method is called.
+    auto h = hints;
+    h.partial_alloc_ok = true;
+    h.is_contiguous = true;
+    blk_count_t nblks_remain = nblks;
+    BlkAllocStatus status;
+
+    do {
+        out_blkids.emplace_back(); // Put an empty MultiBlkId and use that for allocating them
+        BlkId& out_bid = out_blkids.back();
+        status = alloc_contiguous_blks(nblks_remain, h, out_bid);
+
+        auto nblks_this_iter = out_bid.blk_count();
+        nblks_remain = (nblks_remain < nblks_this_iter) ? 0 : (nblks_remain - nblks_this_iter);
+    } while (nblks_remain);
+
+    return status;
+}
+
+BlkAllocStatus VirtualDev::alloc_blks_from_chunk(blk_count_t nblks, blk_alloc_hints const& hints, MultiBlkId& out_blkid,
+                                                 Chunk* chunk) {
 #ifdef _PRERELEASE
     if (auto const fake_status =
             iomgr_flip::instance()->get_test_flip< uint32_t >("blk_allocation_flip", nblks, chunk->vdev_id())) {
         return static_cast< BlkAllocStatus >(fake_status.get());
     }
 #endif
-    static thread_local std::vector< BlkId > chunk_blkid{};
-    chunk_blkid.clear();
-    auto status = chunk->blk_allocator_mutable()->alloc(nblks, hints, chunk_blkid);
-    if (status == BlkAllocStatus::PARTIAL) {
+    auto status = chunk->blk_allocator_mutable()->alloc(nblks, hints, out_blkid);
```

Comment:
BlkAllocator should take MultiBlkId because blk_allocator has functionality and much faster to allocate multiple blks. There is a big logic around varsize_blk_allocation to efficiently do that.

Reply:
sorry my bad, I forgot no overload as we dont do virtual.....